### PR TITLE
Saturn L2 V0

### PR DESCRIPTION
**Pull request by aarshkshah1992**

### CODE REVIEW

**Review by willscott — COMMENTED** [review]
should @bajtos also be reviewing these for integration with station?

**Comment by aarshkshah1992** [comment]
@willscott Don't wanna bother @bajitos with reviewing Golang code and that too one that uses so much of the PL Stack.

 I'll have a separate discussion/doc/long running interaction with him for integrating with Station.

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,42 +1,22 @@
-# Use the latest 2.1 version of CircleCI pipeline process engine.
-# See: https://circleci.com/docs/2.0/configuration-reference
 version: 2.1
-
-# Define a job to be invoked later in a workflow.
-# See: https://circleci.com/docs/2.0/configuration-reference/#jobs
+orbs:
```
[the remark this answers is not in the document]
Comment (aarshkshah1992) [reply .circleci/config.yml:2]: @willscott What is this "standard PL git management  setup" that thou refer to ?

**Comment by willscott** [comment]
him or the L1 devs will know what sort of responses they want to be getting from the l2 interaction so may be better reviewers than me

File: .circleci/config.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (willscott) [reply .circleci/config.yml:2]: https://github.com/protocol/.github/

**Comment by bajtos** [comment]
@aarshkshah1992 I am happy to learn more about Go :) I quickly skimmed through your code and have one question - how is this new code and the new HTTP server integrating with the existing `main` binary? Do we want two HTTP servers listening on different ports (one for web, another for Saturn)?

**Comment by willscott** [comment]
My understanding is that the golang code here will be a libp2p node, and will make connections outbound to L1's.
The HTTP server here gets requests funneled through that libp2p connection - the actual network request is a long lived L2->L1 to handle cases where L2's are behind NATs

File: carstore/gateway_api.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/filecoin-project/dagstore/mount"
+	cid "github.com/ipfs/go-cid"
+)
+
+var (
+	defaultURL = "https://ipfs.io/api/v0/dag/export"
```
Comment (lidel) [inline carstore/gateway_api.go:14]: nit:  go-ipfs 0.13+ allows you to replace legacy RPC API at  `/api/v0/dag/export` with implementation-agnostic  Gateway request `GET https://ipfs.io/ipfs/{cid}?format=car`   (or `GET https://ipfs.io/ipfs/{cid}` with  `Accept: application/vnd.ipld.car`)

This way you will be able to swap gateway endpoint and its  implementation in the future + benefit from planned [selector support](https://github.com/ipfs/go-ipfs/issues/8769).

File: carserver/server_test.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	temp := t.TempDir()
+	ctx := context.Background()
+
+	// create the getway api with a test http server
+	root, contents, svc := testutils.GetTestServerFor(t, path)
+	defer svc.Close()
+	gwAPI := carstore.NewGatewayAPI(svc.URL)
+	lg := logs.NewSaturnLogger()
+	cfg := carstore.Config{}
+	cs, err := carstore.New(temp, gwAPI, cfg, lg)
```
Comment (bajtos) [inline carserver/server_test.go:50]: When running inside Filecoin Station, we need Saturn L2 node to store all cache files in the directory provided by the Station. (Typically, this directory is OS-specific. See https://github.com/filecoin-project/saturn-l2/issues/36 and https://github.com/ipfs/ipfs-desktop/issues/1656 for more details.)

I am proposing to introduce a new env var, e.g. `CACHE_DIR` or `SATURN_CACHEDIR`:
- Filecoin Station sets `CACHE_DIR` env var, similarly to how it sets `FIL_WALLET_ADDRESS`
- Saturn L2 Node uses this value as the root dir for all caches

To make it easier to run saturn-l2 on its own, we can introduce a sensible default when `CACHE_DIR` is not provided. I don't think `t.TempDir()` is a good one - IMHO, we should preserve cached data across node restarts?

Comment (aarshkshah1992) [reply carserver/server_test.go]: Hey I have introduced an ENV VAR called ROOT_DIR.

Comment (aarshkshah1992) [reply carserver/server_test.go]: I haven't added a sensible default yet but we can add it down the line if need be. Should be easy.

**Comment by aarshkshah1992** [comment]
@bajtos  -> Yes good suggestion, I have moved all binaries to cmd/.

**Pull request merged**